File: array/cpp/BinarySearch.cpp
// Binary Search in C++

#include <iostream>
using namespace std;

int binarySearch(int start, int end,int arr[], int target) {
    if (end<start){
        return -1;
    }

    int mid = (start+end)/2;
        
    if (arr[mid] == target){
        return mid;
    }

    if (arr[mid]<target){
        return binarySearch(mid+1, end, arr, target);
    }
        
    return binarySearch(start, mid-1, arr, target);
}

int main(void) {
    int arr[] = {1,2,3,4,5,6};
    int target = 4;
    int size = sizeof(arr) / sizeof(arr[0]);
    int position = binarySearch(0, size - 1,arr, target);
    cout << "Position of " << target << " is " << position;
}

File: array/java/arvind/BinarySearch.java
package arvind;

public class BinarySearch {

    public int find(int start, int end, int[] arr, int target) {
        if (end < start) {
            return -1;
        }
        int mid = end - (end - start) / 2;

        if (arr[mid] == target) {
            return mid;
        }
        if (arr[mid] > target) {
            return find(start, mid - 1, arr, target);
        } else {
            return find(mid + 1, end, arr, target);
        }
    }

    public int find(int[] arr, int target) {
        int start = 0, end = arr.length - 1;


        while (end >= start) {
            int mid = end - (end - start) / 2;
            if (arr[mid] == target) {
                return mid;
            } else {
                if (arr[mid] > target) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
        }


        return -1;
    }

    public static void main(String[] args) {
        BinarySearch binarySearch = new BinarySearch();
        int[] arr = new int[] {2, 3, 5, 9, 10, 11};
        int target = 2;
        int result = binarySearch.find(0, arr.length-1 , arr, target);
        System.out.println("Binary search output is "+ result);
        int result2 = binarySearch.find(arr, target);
        System.out.println("Binary search output is "+ result2);
    }
}
